contracts/allocators/OnsenAllocatorV2.sol

```diff
@@ -0,0 +1,268 @@
+        } else {
+            return 0;
+        }
+    }
+
+    function rewardTokens() public view override returns (IERC20[] memory) {
+        IERC20[] memory rewards = new IERC20[](2);
+        rewards[0] = IERC20(sushi);
+        rewards[1] = IERC20(xSushi);
+        return rewards;
+    }
+
+    function utilityTokens() public view override returns (IERC20[] memory) {
+        IERC20[] memory utility = new IERC20[](1);
+        utility[0] = IERC20(sushi);
+        return utility;
+    }
+
+    function name() external pure override returns (string memory) {
+        return "OnsenAllocator";
+    }
+
+    function setTreasury(address treasuryAddress) external onlyGuardian {
+        treasury = treasuryAddress;
+    }
+
+    function setSushi(address sushiAddress) external onlyGuardian {
+        sushi = sushiAddress;
+    }
+
+    function setXSushi(address xSushiAddress) external onlyGuardian {
+        xSushi = xSushiAddress;
+    }
+
+    function setMasterChefAddress(address masterChefAddress) external onlyGuardian {
+        masterChef = masterChefAddress;
+    }
+
+    /// ========== INTERNAL FUNCTIONS ==========
+
+    /**
+     * @notice stake sushi rewards
+     */
+    function _stakeSushi() internal {
+        uint256 balance = IERC20(sushi).balanceOf(address(this));
+        if (balance > 0) {
+            IERC20(sushi).approve(xSushi, balance);
+            ISushiBar(xSushi).enter(balance); // stake sushi
+        }
+    }
+
+    /**
+     * @notice unStake sushi rewards
+     */
+    function _unStakeSushi() internal {
+        uint256 balance = IERC20(xSushi).balanceOf(address(this));
+        if (balance > 0) {
+            ISushiBar(xSushi).leave(balance); // unstake $xSUSHI
+        }
+    }
+
+    /**
+     *  @notice pending $SUSHI rewards
+     *  @return uint
+     */
+    function _pendingSushi(uint256 pid) internal view returns (uint256) {
+        return IMasterChef(masterChef).pendingSushi(pid, address(this));
+    }
+
+    /**
+     * @notice Find out pool Id from Onsen based on the LP token
+     * according to Onsen documentation there shouldn't be more than pool by LPToken, so we are going to use the first ocurrence of the LPToken in the Onsen pools
+     * @return [bool, uint256], first value shows if value was found, second value the id for the pool
+     */
+    function _findPoolByLP(address LPToken) internal view returns (bool, uint256) {
+        /// Check if the id is already stored in the local variable
+        PoolData memory pd = _lpToOnsenId[LPToken];
+
+        /// If the pool is active return the id else search for the pool id
+        if (pd.poolActive) {
+            return (pd.poolActive, pd.id);
+        } else {
+            /// Call the poolLength function from sushi masterchef v2 and compare against the LP token passed as parameter
+            uint256 poolsLength = IMasterChef(masterChef).poolLength();
+
+            for (uint256 i; i < poolsLength; i++) {
```

Comment:
Poollength of masterchef is 355 i think unless im looking at wrong contract. i think this can be done offchain. We can figure out the id mappings offchain using view functions on masterchef and insert them in with a setter function. This would work but im worried it will use alot more gas than we need to.

Reply:
Added the following 2 functions:

function setLPTokenOnsenPoolId(address lpToken, uint256 onsenPoolId) external onlyGuardian 
function getLPTokenOnsenPoolId(address lpToken) external view returns (bool, uint256)

So we can add the Onsen Pool Id for the LP Tokens ahead of time and also check if the LP token was already added, worst-case scenario would be that the token is not added and it will go search for it based on the Onsen Pool Length.

@Paul4912 Please let me know if this is fine or if we need to be able to add all Onsen Pool Ids at once?

Reply:
Yea that's how id imagine we'd do it. 

Idk how many onsen farms there are to need a batch add.

Reply:
@PARABRAHMAN0 @ind-igo Do you have any idea about how many Onsen Pools are used? Or do you know where I can check this? 
We would like to know if a batch function to add Onsen pools is needed or not?